zookeeper-client/zookeeper-client-c/src/zookeeper.c

```diff
@@ -3961,6 +3961,22 @@ static int Request_path_watch_init(zhandle_t *zh, int mode,
 /*---------------------------------------------------------------------------*
  * ASYNC API
  *---------------------------------------------------------------------------*/
+static int nonblocking_send(zhandle_t *zh, int rc)
+{
+    if (rc < 0) {
+        rc = ZMARSHALLINGERROR;
+    } else {
+        rc = adaptor_send_queue(zh, 0);
```

Comment:
(sorry I have one last question, I could have noticed it before).

in the previous version of this code we were basically always calling `adaptor_send_queue`, now in case of previous error (rc < 0) we are not calling that function anymore.

can this have a behavioural impact ?

Reply:
Thanks, this is a good question.  I have reorganized the code to minimize the change; now adaptor_send_queue() will always be called.

If an error is returned from adaptor_send_queue(), check if we are in a connected state (`is_connected`) and only then, close the socket and set state to `ZOO_NOTCONNECTED_STATE`.

Also: removed the replicated comment and collapsed the ZMARSHALLINGERROR check in `aremove_watches`.